core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -1116,62 +708,89 @@ protected boolean shouldExecuteReplication(Settings settings) {
         return IndexMetaData.isIndexUsingShadowReplicas(settings) == false;
     }
 
-    interface IndexShardReference extends Releasable {
-        boolean isRelocated();
-
-        void failShard(String reason, @Nullable Throwable e);
-
-        ShardRouting routingEntry();
-
-        /** returns the primary term of the current operation */
-        long opPrimaryTerm();
-    }
-
-    static final class IndexShardReferenceImpl implements IndexShardReference {
+    class PrimaryShardReference implements ReplicatedOperation.Primary<Request, ReplicaRequest, Response>, Releasable {
 
         private final IndexShard indexShard;
         private final Releasable operationLock;
 
-        private IndexShardReferenceImpl(IndexShard indexShard, long primaryTerm) {
+        PrimaryShardReference(IndexShard indexShard, Releasable operationLock) {
             this.indexShard = indexShard;
-            if (primaryTerm < 0) {
-                operationLock = indexShard.acquirePrimaryOperationLock();
-            } else {
-                operationLock = indexShard.acquireReplicaOperationLock(primaryTerm);
-            }
+            this.operationLock = operationLock;
         }
 
-        static IndexShardReferenceImpl createOnPrimary(IndexShard indexShard) {
-            return new IndexShardReferenceImpl(indexShard, -1);
+        @Override
+        public void close() {
+            operationLock.close();
         }
 
-        static IndexShardReferenceImpl createOnReplica(IndexShard indexShard, long primaryTerm) {
-            return new IndexShardReferenceImpl(indexShard, primaryTerm);
+        public boolean isRelocated() {
+            return indexShard.state() == IndexShardState.RELOCATED;
         }
 
         @Override
-        public void close() {
-            operationLock.close();
+        public void failShard(String reason, Throwable e) {
+            try {
+                indexShard.failShard(reason, e);
+            } catch (Throwable suppressed) {
+                e.addSuppressed(suppressed);
```

Comment:
we dont really swallow it but add it to the failure reason. We're protecting here against AlreadyClosedException that the shard may throw (among other potential issues with double failures).